crates/forge-runner/src/runtime.rs

```diff
@@ -0,0 +1,204 @@
+        if let Some(Hint::Starknet(StarknetHint::Cheatcode {
+            selector,
+            input_start,
+            input_end,
+            output_start,
+            output_end,
+        })) = maybe_extended_hint
+        {
+            // Parse the selector.
+            let selector = &selector.value.to_bytes_be().1;
+            let selector = std::str::from_utf8(selector).map_err(|_| {
+                CustomHint(Box::from(
+                    "Failed to parse the  cheatcode selector".to_string(),
+                ))
+            })?;
+
+            // Extract the inputs.
+            let input_start = extract_relocatable(vm, input_start)?;
+            let input_end = extract_relocatable(vm, input_end)?;
+            let inputs = vm_get_range(vm, input_start, input_end)
+                .map_err(|_| CustomHint(Box::from("Failed to read input data".to_string())))?;
+
+            if let CheatcodeHadlingResult::Result(()) =
+                self.0
+                    .handle_cheatcode(selector, inputs, vm, output_start, output_end)?
+            {
+                return Ok(());
+            }
+        }
+
+        if let Some(Hint::Starknet(StarknetHint::SystemCall { system })) = maybe_extended_hint {
+            // TODO move selector parsing logic here
+            if let SyscallHandlingResult::Result(()) = self.0.override_system_call(system, vm)? {
+                return Ok(());
+            }
+        }
+        self.0
+            .get_extended_runtime_mut()
+            .execute_hint(vm, exec_scopes, hint_data, constants)
+    }
+
+    fn compile_hint(
+        &self,
+        hint_code: &str,
+        ap_tracking_data: &ApTracking,
+        reference_ids: &HashMap<String, usize>,
+        references: &[HintReference],
+    ) -> Result<Box<dyn Any>, VirtualMachineError> {
+        self.0.get_extended_runtime().compile_hint(
+            hint_code,
+            ap_tracking_data,
+            reference_ids,
+            references,
+        )
+    }
+}
+
+impl<Handler, Runtime: HintProcessor> ResourceTracker
+    for ExtendedRuntime<RuntimeExtension<Handler, Runtime>>
+{
+    fn consumed(&self) -> bool {
+        self.0.extended_runtime.consumed()
+    }
+
+    fn consume_step(&mut self) {
+        self.0.extended_runtime.consume_step();
+    }
+
+    fn get_n_steps(&self) -> Option<usize> {
+        self.0.extended_runtime.get_n_steps()
+    }
+
+    fn run_resources(&self) -> &RunResources {
+        self.0.extended_runtime.run_resources()
+    }
+}
+
+#[allow(dead_code)]
+#[derive(Debug)]
+pub enum SyscallHandlingResult {
+    Forward,
+    Result(()),
+}
+
+#[allow(dead_code)]
+#[derive(Debug)]
+pub enum CheatcodeHadlingResult {
+    Forward,
+    Result(()), // TODO now use buffer later rewrite to return vector
+}
+
+pub trait ExtensionLogic {
+    type Runtime: HintProcessorLogic;
+
+    fn get_extended_runtime_mut(&mut self) -> &mut Self::Runtime;
+
+    fn get_extended_runtime(&self) -> &Self::Runtime;
```

Comment:
I don't get the point of these

Reply:
Those are required to ensure type of Runtime has `HintProcessorLogic` implemented

Reply:
It looks more like a hack, I don't like it tbh

Reply:
I will take a look if this can be simplified, but keep in mind

> It looks more like a hack, I don't like it tbh
is not very high quality feedback on somebodys work. Quality feedback includes the understanding of a tradeoffs and offers a suggestion how to improve things.